README.md
<p align="center">
  <img alt="angular-material-extensions's logo"
   height="256px" width="256px" style="text-align: center;"
   src="https://cdn.jsdelivr.net/gh/angular-material-extensions/select-country@master/assets/angular-material-extensions-logo.svg">
</p>

# @angular-material-extensions/select-country - Angular Material component that allow users to select a country or nationality with an autocomplete feature

[![npm version](https://badge.fury.io/js/%40angular-material-extensions%2Fselect-country.svg)](https://badge.fury.io/js/%40angular-material-extensions%2Fselect-country)
[![npm demo](https://img.shields.io/badge/demo-online-ed1c46.svg)](https://angular-material-extensions.github.io/select-country)
[![docs: typedoc](https://img.shields.io/badge/docs-typedoc-4D0080.svg)](https://angular-material-extensions.github.io/select-country/doc/index.html)
[![Join the chat at https://gitter.im/angular-material-extensions/Lobby](https://badges.gitter.im/angular-material-extensions/Lobby.svg)](https://gitter.im/angular-material-extensions/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://travis-ci.org/angular-material-extensions/select-country.svg?branch=master)](https://travis-ci.org/angular-material-extensions/select-country)
[![codecov](https://codecov.io/gh/angular-material-extensions/select-country/branch/master/graph/badge.svg)](https://codecov.io/gh/angular-material-extensions/select-country)
[![Greenkeeper Badge](https://badges.greenkeeper.io/angular-material-extensions/select-country.svg)](https://greenkeeper.io/)
[![license](https://img.shields.io/github/license/angular-material-extensions/select-country.svg?style=flat-square)](https://github.com/angular-material-extensions/select-country/blob/master/LICENSE)

<p align="center">
  <img alt="@angular-material-extensions/select-country demonstration" style="text-align: center;"
   src="https://raw.githubusercontent.com/angular-material-extensions/select-country/HEAD/assets/v0.2.0/select-country.gif">
</p>

<p align="center">
  <img alt="@angular-material-extensions/select-country demonstration" style="text-align: center;"
   src="https://raw.githubusercontent.com/angular-material-extensions/select-country/HEAD/assets/v0.2.0/select-country.png">
</p>

## Built by and for developers :heart:
Do you have any question or suggestion ? Please do not hesitate to contact us!
Alternatively, provide a PR | open an appropriate issue [here](https://github.com/angular-material-extensions/select-country/issues)

If you like this project, support [angular-material-extensions](https://github.com/angular-material-extensions)
by starring :star: and sharing it :loudspeaker:

## Table of Contents
- [Demo](#demo)
- [Components](#components)
- [Dependencies](#dependencies)
- [Installation](#installation)
- [API](#api)
- [Usage](#usage)
- [Run Demo App Locally](#run-demo-app-locally)
- [Other Angular Libraries](#other-angular-libraries)
- [Support](#support)
- [License](#license)

<a name="demo"/>

## [Demo](https://angular-material-extensions.github.io/select-country)

View all the directives and components in action at [https://angular-material-extensions.github.io/select-country](https://angular-material-extensions.github.io/select-country)

<a name="components"/>

## Library's components
- `<mat-select-country>` used to display the main component

---

<a name="dependencies"/>

## Dependencies
* [Angular](https://angular.io) developed and tested with `17.x`

---

<a name="installation"/>

##  [Installation](https://angular-material-extensions.github.io/select-country/getting-started)

## 1. Install via *ng add*. (Recommended)

If Angular Material Design is not setup, just run `ng add @angular/material` [learn more](https://material.angular.io/guide/getting-started)

Now add the library via the `angular schematics`
```shell
ng add @angular-material-extensions/select-country
```

## 2. Install via *npm*. (Alternative)

Install peer dependencies
```shell
npm i svg-country-flags -s
```

then update your `angular.json` like below (svg-country-flags)

```json
"assets": [
              "src/favicon.ico",
              "src/assets",
              {
                "glob": "**/*",
                "input": "./node_modules/svg-country-flags/svg",
                "output": "src/assets/svg-country-flags/svg"
              }
            ],
```

Now install `@angular-material-extensions/select-country` via:

```shell
npm install --save @angular-material-extensions/select-country
```

### Import the library

If you installed the library via angular schematics, you can skip this step

Once installed you need to import the main module and the `HttpClientModule`:

```js
import { MatSelectCountryModule } from "@angular-material-extensions/select-country";
```

```typescript
import { MatSelectCountryModule } from '@angular-material-extensions/select-country';
import {HttpClientModule} from '@angular/common/http';

@NgModule({
  declarations: [AppComponent, ...],
  imports: [
              MatSelectCountryModule.forRoot('de'), // you can use 'br' | 'de' | 'en' | 'es' | 'fr' | 'hr' | 'hu' | 'it' | 'nl' | 'pt' --> MatSelectCountrySupportedLanguages
             HttpClientModule, ...],
  bootstrap: [AppComponent]
})
export class AppModule {
}
```

in other modules

```typescript
import { MatSelectCountryModule } from '@angular-material-extensions/select-country';
import {HttpClientModule} from '@angular/common/http';

@NgModule({
  declarations: [AppComponent, ...],
  imports: [MatSelectCountryModule, HttpClientModule, ...],
  bootstrap: [AppComponent]
})
export class AppModule {
}
```

Other modules in your application like for lazy loading import `MatSelectCountryModule` into your feature module:

<a name="api"/>

## API

### `<mat-select-country>` used to display the main component - [see the demo examples](https://angular-material-extensions.github.io/select-country/examples)

| option            |    bind    |           type           |             default             | description                                                                                        |
| :---------------- | :--------: | :----------------------: | :-----------------------------: | :------------------------------------------------------------------------------------------------- |
| value             | `Input()`  |        `Country`         |                -                | the selected country (pass an empty Country object with alpha2 code populated to do alpha2 lookup) |
| appearance        | `Input()`  | `MatFormFieldAppearance` |                -                | Possible appearance styles for `mat-form-field` ('legacy', 'standard', 'fill' or 'outline')        |
| countries         | `Input()`  |       `Country[]`        | All countries stored in the lib | Countries that should be loaded - predefine the countries that you only need!                      |
| label             | `Input()`  |        `boolean`         |                -                | `mat-form-field` label's text                                                                      |
| itemsLoadSize     | `Input()`  |         `number`         |                -                | the number of countries that should be fetched --> improves the performance                        |
| placeHolder       | `Input()`  |        `boolean`         |                -                | input placeholder text                                                                             |
| disabled          | `Input()`  |        `boolean`         |                -                | Whether the component is disabled                                                                  |
| loading           | `Input()`  |        `boolean`         |                -                | Whether the component is loading                                                                   |
| nullable          | `Input()`  |        `boolean`         |                -                | Whether the component is able to emit `null`                                                       |
| readonly          | `Input()`  |        `boolean`         |                -                | Whether the component is read only                                                                 |
| tabIndex          | `Input()`  |    `number \| string`    |                -                | Whether the component can be focused, and where it participates in sequential keyboard navigation  |
| showCallingCode   | `Input()`  |        `boolean`         |              false              | Whether the component to show the country's calling code in the label and selection                |
| class             | `Input()`  |         `string`         |                -                | Class attribute apply style to input text or validation ignore (optional)                          |
| language          | `Input()`  |         `string`         |                -                | the language, if not specified MatSelectCountryModule.forRoot('XX') will be used (optional)        |
| name              | `Input()`  |         `string`         |             'country'           | the attribute name of the input element                                                            |
| autocomplete      | `Input()`  |         `string`         |                -                | the attribute autocomplete of the input element, to avoid suggestion of some browsers put 'no'     |
| onCountrySelected | `Output()` | `EventEmitter<Country>`  |                -                | emits the selected country as object (see the interface below)                                     |

```typescript
interface Country {
  name: string;
  alpha2Code: string;
  alpha3Code: string;
  numericCode: string;
  callingCode: string;
}
```


<a name="usage"/>

## [Usage](https://angular-material-extensions.github.io/select-country)

add the `<mat-select-country>` element to your template:

```html
<mat-select-country>
</mat-select-country>
```

<p align="center">
  <img alt="@angular-material-extensions/select-country demonstration" style="text-align: center;"
   src="https://raw.githubusercontent.com/angular-material-extensions/select-country/HEAD/assets/v0.2.0/example_full.png">
</p>

<p align="center">
  <img alt="@angular-material-extensions/select-country demonstration" style="text-align: center;"
   src="https://raw.githubusercontent.com/angular-material-extensions/select-country/HEAD/assets/v0.2.0/example3.png">
</p>

<p align="center">
  <img alt="@angular-material-extensions/select-country demonstration" style="text-align: center;"
   src="https://raw.githubusercontent.com/angular-material-extensions/select-country/HEAD/assets/v0.2.0/example2.png">
</p>


#### Use the library with reactive forms

### option 1 with Reactive Forms

```html
<mat-select-country appearance="outline"
                    label="Country"
                    [formControl]="countryFormControl"
                    (onCountrySelected)="onCountrySelected($event)">
</mat-select-country>
```

### option 2 with Reactive Forms

```html
<form [formGroup]="countryFormGroup">
    <div fxLayoutAlign="center">
      <mat-select-country appearance="outline"
                          label="Country"
                          class="className"
                          formControlName="country"
                          (onCountrySelected)="onCountrySelected($event)">
      </mat-select-country>
    </div>
  </form>
```

```typescript
import {Component,OnInit} from '@angular/core';
import {FormControl} from '@angular/forms';

import {Country} from '@angular-material-extensions/select-country';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent implements OnInit{
  title = 'select-country';

  countryFormControl = new FormControl();
  countryFormGroup: FormGroup;

  constructor(private formBuilder: FormBuilder) {
    angulartics2GoogleAnalytics.startTracking();
  }

  ngOnInit(): void {

    this.countryFormGroup = this.formBuilder.group({
      country: []
    });

    this.countryFormGroup.get('country').valueChanges
.subscribe(country => console
.log('this.countryFormGroup.get("country").valueChanges', country));

    this.countryFormControl.valueChanges
.subscribe(country => console
.log('this.countryFormControl.valueChanges', country));
  }


  onCountrySelected($event: Country) {
    console.log($event);
  }
}

```


### Predefine your countries to load

```html
<mat-select-country appearance="outline"
                    label="Country"
                    [countries]="predefinedCountries"
                    (onCountrySelected)="onCountrySelected($event)">
</mat-select-country>
```

```typescript
import {Country} from '@angular-material-extensions/select-country';

predefinedCountries: Country[] = [
  {
    name: 'Germany',
    alpha2Code: 'DE',
    alpha3Code: 'DEU',
    numericCode: '276',
    callingCode: '+49'
  },
  {
    name: 'Greece',
    alpha2Code: 'GR',
    alpha3Code: 'GRC',
    numericCode: '300',
    callingCode: '+30'
  },
  {
    name: 'France',
    alpha2Code: 'FR',
    alpha3Code: 'FRA',
    numericCode: '250',
    callingCode: '+33'
  },
  {
    name: 'Belgium',
    alpha2Code: 'BE',
    alpha3Code: 'BEL',
    numericCode: '056',
    callingCode: '+32'
  }
];
```

Result:

<p align="center">
  <img alt="@angular-material-extensions/select-country demonstration" style="text-align: center;"
   src="https://raw.githubusercontent.com/angular-material-extensions/select-country/HEAD/assets/v2.1.0/predefined.png">
</p>


### Improve performance - use the `itemsLoadSize` property

```html
<mat-select-country appearance="outline"
                    label="Country"
                    [itemsLoadSize]="5">
</mat-select-country>
```

only 5 countries will be fetched!

### Change language dynamically - use the `language` property
```html
<mat-select-country appearance="outline"
                    label="Country"
                    [language]="languageSelected"
                    (onCountrySelected)="onCountrySelected($event)">
</mat-select-country>
```


<a name="run-demo-app-locally"/>

###  Run Demo App Locally


Build the library

```bash
$ npm run build:lib
```

Serve the demo app

```bash
$ npm start
```



## Other Angular Libraries
- [ngx-auth-firebaseui](https://github.com/AnthonyNahas/ngx-auth-firebaseui)
- [ngx-linkifyjs](https://github.com/AnthonyNahas/ngx-linkifyjs)
- [@angular-material-extensions/password-strength](https://github.com/angular-material-extensions/password-strength)
- [@angular-material-extensions/google-maps-autocomplete](https://github.com/angular-material-extensions/google-maps-autocomplete)
- [@angular-material-extensions/link-preview](https://github.com/angular-material-extensions/link-preview)
- [@angular-material-extensions/fab-menu](https://github.com/angular-material-extensions/fab-menu)
- [@angular-material-extensions/pages](https://github.com/angular-material-extensions/pages)
- [@angular-material-extensions/contacts](https://github.com/angular-material-extensions/contacts)
---

<a name="who_is_using_this_lib"/>

## Who is using this library?

[nahaus.de](https:/nahaus.de) - Digital and Automated Real Estate Management

To put your project here, please drop an appropriate PR

-----

<a name="support"/>

## Support
+ Drop an email to: [Anthony Nahas](mailto:anthony.na@hotmail.de)
+ or open an appropriate [issue](https://github.com/angular-material-extensions/select-country/issues)
+ let us chat on [Gitter](https://gitter.im/angular-material-extensions/Lobby)

 Built by and for developers :heart: we will help you :punch:

---

![jetbrains logo](https://raw.githubusercontent.com/angular-material-extensions/select-country/HEAD/assets/jetbrains-variant-4_logos/jetbrains-variant-4.png)

This project is supported by [jetbrains](https://www.jetbrains.com/) with 1 ALL PRODUCTS PACK OS LICENSE incl. [webstorm](https://www.jetbrains.com/webstorm)

---

<a name="license"/>

## License

Copyright (c) 2020-2024 [Anthony Nahas](https://github.com/AnthonyNahas). Licensed under the MIT License (MIT) <p align="center">
                                                                                                            <img alt="angular-material-extensions's logo"
                                                                                                             height="92px" width="92px" style="text-align: center;"
                                                                                                             src="https://cdn.jsdelivr.net/gh/angular-material-extensions/select-country@master/assets/badge_made-in-germany.svg">
                                                                                                          </p>


CHANGELOG.md


# [17.0.0](https://github.com/angular-material-extensions/select-country/compare/16.0.0...17.0.0) (2023-11-19)


### Bug Fixes

* **project:** minor ([8f393fc](https://github.com/angular-material-extensions/select-country/commit/8f393fc6e69e901cb10b38caa52423ed802a2497))
* **project:** upgraded angular material to v17 ([98e426d](https://github.com/angular-material-extensions/select-country/commit/98e426d4e64c6fadd5a59486308fd94b8a95587c))
* **project:** upgraded angular to v17 ([1aff047](https://github.com/angular-material-extensions/select-country/commit/1aff0471c0aca03b625f290c4a58168c561e6c25))
* **project:** upgraded release it ([6826638](https://github.com/angular-material-extensions/select-country/commit/6826638ce14a5fef9ef0ba4af622c8578775dcb1))

# [16.0.0](https://github.com/angular-material-extensions/select-country/compare/8.1.0...16.0.0) (2023-11-19)


### Bug Fixes

* **project:** minor ([a04507f](https://github.com/angular-material-extensions/select-country/commit/a04507fb45e7de3cea3197aac110a9ac9e8bff8d))
* **project:** updated deps ([fbfba2c](https://github.com/angular-material-extensions/select-country/commit/fbfba2c8740177702bcd7895a2945484d711427a))
* **project:** updated release it deps ([2d4003b](https://github.com/angular-material-extensions/select-country/commit/2d4003b7fee0835a4687e4d956dbe0e0481ecaf3))
* **project:** upgraded angular to v16 ([825f87d](https://github.com/angular-material-extensions/select-country/commit/825f87d5d9f1ece34f4af35a16af44889d41d574))
* **project:** upgraded angular universal to v16 ([0193e83](https://github.com/angular-material-extensions/select-country/commit/0193e83772656ccf3a03e3758fc1892c2551add7))
* **project:** upgraded flex layout ([e0f885a](https://github.com/angular-material-extensions/select-country/commit/e0f885a415aeda8b36e63469d3c64357dd0192b7))

# [8.1.0](https://github.com/angular-material-extensions/select-country/compare/8.0.0...8.1.0) (2023-01-14)

# [8.0.0](https://github.com/angular-material-extensions/select-country/compare/7.1.0...8.0.0) (2022-11-30)


### Bug Fixes

* **lib:** adapt lib to v15 ([4c26efe](https://github.com/angular-material-extensions/select-country/commit/4c26efee5e68a8ee6da2ceb50a4104ac8a1314e7))
* **lib:** upgraded angular core and cli to v15 ([b7ac710](https://github.com/angular-material-extensions/select-country/commit/b7ac710a84b1c9a6108f9bd3dec9fbf28590db4d))
* **lib:** upgraded angular material to v15 ([b494af5](https://github.com/angular-material-extensions/select-country/commit/b494af5d3bd6c7a99fe7fff11154c07106a6a6ec))
* **lib:** upgraded angular peer deps ([82f3fcc](https://github.com/angular-material-extensions/select-country/commit/82f3fccde7f74191a158a89a4e1f841fde09a6af))
* **lib:** upgraded angular universal ([427c32b](https://github.com/angular-material-extensions/select-country/commit/427c32b72dd71b352cb71a44425c45e7179544e5))

# [7.1.0](https://github.com/angular-material-extensions/select-country/compare/7.0.1...7.1.0) (2022-11-18)

## [7.0.1](https://github.com/angular-material-extensions/select-country/compare/7.0.0...7.0.1) (2022-07-04)


### Bug Fixes

* **lib:** minor ([f7a6bc6](https://github.com/angular-material-extensions/select-country/commit/f7a6bc67918017cee9892cb4c4c8d29188d783d9))
* **lib:** minor ([7d5640f](https://github.com/angular-material-extensions/select-country/commit/7d5640ff50d1dffca2150234cc783b912a356a30))
* **lib:** upgeaded angular core and cli ([89e48a8](https://github.com/angular-material-extensions/select-country/commit/89e48a8560e46882dd3c6e792512546c820a2d40))
* **lib:** upgeaded angular material ([4c4a9ec](https://github.com/angular-material-extensions/select-country/commit/4c4a9ec6872b737697e09592300a3cb884d10958))
* **lib:** upgeaded angular universal ([1fad261](https://github.com/angular-material-extensions/select-country/commit/1fad261614cab1790e885368f3765f705f0c1288))

* Merge remote-tracking branch 'origin/master' (dca4528)
* fix(lib): minor (7d5640f)
* Create FUNDING.yml (05aba23)

# [7.0.0](https://github.com/angular-material-extensions/select-country/compare/6.0.1...7.0.0) (2022-06-21)


### Bug Fixes

* **lib:** minor ([e7a7c5c](https://github.com/angular-material-extensions/select-country/commit/e7a7c5c026694d9fdd97f74346d2c5f92b3360a0))
* **lib:** upgraded angular deps ([ae51ea4](https://github.com/angular-material-extensions/select-country/commit/ae51ea49b2457f7656ac41beeb9ede06fbee7b52))
* **lib:** upgraded angular material deps ([b53a4e2](https://github.com/angular-material-extensions/select-country/commit/b53a4e28e0f3332a89f3782b4e7c0eec8433decd))
* **lib:** upgraded angular universal deps ([7e9ad14](https://github.com/angular-material-extensions/select-country/commit/7e9ad14e1308a51b387b3cbcffd84a5a53a0265d))

## [6.0.1](https://github.com/angular-material-extensions/select-country/compare/6.0.0...6.0.1) (2022-01-05)


### Bug Fixes

* **docs:** updated the readme ([da45022](https://github.com/angular-material-extensions/select-country/commit/da45022336dc2bbd704e6853477fdf653e663abf))

# [6.0.0](https://github.com/angular-material-extensions/select-country/compare/5.1.0...6.0.0) (2022-01-05)


### Bug Fixes

* **lib:** minor ([91635af](https://github.com/angular-material-extensions/select-country/commit/91635af8e58af2251af1b77ca7e7c62c8eb7287c))
* **project:** upgraded angular material to v13 ([feb96cf](https://github.com/angular-material-extensions/select-country/commit/feb96cfa09e14705c3e0a2ed86ecd2ed872fbf62))
* **project:** upgraded angular material to v13 ([c12f89d](https://github.com/angular-material-extensions/select-country/commit/c12f89d8389d055ac505a50e5df99c556558d08a))
* **project:** upgraded angular material to v13 ([65d261a](https://github.com/angular-material-extensions/select-country/commit/65d261ade5589f4f7b919eeae79e3cad2962a62c))
* **project:** upgraded angular material to v13 ([2d555fe](https://github.com/angular-material-extensions/select-country/commit/2d555fe98091a13aedeb5cfefa579c02083215e0))

src/app/app.component.ts
import {Component, OnInit} from "@angular/core";
import {FormControl, UntypedFormBuilder, UntypedFormControl, UntypedFormGroup, Validators,} from "@angular/forms";

import {Country} from "@angular-material-extensions/select-country";
import {Angulartics2GoogleAnalytics} from 'angulartics2';
import {GERMANY_COUNTRY} from "./examples/contants";

@Component({
  selector: "app-root",
  templateUrl: "./app.component.html",
  styleUrls: ["./app.component.scss"],
})
export class AppComponent implements OnInit {
  title = "select-country";

  defaultValue = {...GERMANY_COUNTRY};
  countryFormControl = new UntypedFormControl();
  countryRequiredFormControl = new UntypedFormControl(null, [
    Validators.required,
  ]);
  countryFormGroup: UntypedFormGroup;

  languageSelected = "es";
  showCallingCode = "false";

  countryMandatoryFormControl = new FormControl(this.defaultValue.name, [
    Validators.required,
  ]);

  predefinedCountries: Country[] = [
    {
      name: "Germany",
      alpha2Code: "DE",
      alpha3Code: "DEU",
      numericCode: "276",
      callingCode: "+49",
    },
    {
      name: "Greece",
      alpha2Code: "GR",
      alpha3Code: "GRC",
      numericCode: "300",
      callingCode: "+30",
    },
    {
      name: "France",
      alpha2Code: "FR",
      alpha3Code: "FRA",
      numericCode: "250",
      callingCode: "+33",
    },
    {
      name: "Belgium",
      alpha2Code: "BE",
      alpha3Code: "BEL",
      numericCode: "056",
      callingCode: "+32",
    },
  ];

  constructor(
    angulartics2GoogleAnalytics: Angulartics2GoogleAnalytics,
    private formBuilder: UntypedFormBuilder
  ) {
    angulartics2GoogleAnalytics.startTracking();
  }

  ngOnInit(): void {
    this.countryFormGroup = this.formBuilder.group({
      country: [{...this.defaultValue}],
    });

    this.countryMandatoryFormControl.valueChanges.subscribe((country) => {
      console.log(
        "this.countryMandatoryFormControl.get('country').valueChanges",
        country
      );
    });

    this.countryFormGroup
      .get("country")
      .valueChanges.subscribe((country) =>
      console.log(
        "this.countryFormGroup.get('country').valueChanges",
        country
      )
    );

    this.countryFormControl.valueChanges.subscribe((country) =>
      console.log("this.countryFormControl.valueChanges", country)
    );
  }

  onCountrySelected($event: Country): void {
    console.log("onCountrySelected", $event);
  }
}
